### ISSUE TITLE
Support build tool integration

### ISSUE THREAD

**Issue by alfaix** [alfaix/neotest-gtest#12]
Ideally, we would like to automatically discover executables for tests with Bazel, CMake, and Conan.

For CMake [cmake-tools.nvim](https://github.com/Civitasv/cmake-tools.nvim) may be useful.

**Comment by alejandroclaro** [comment]
Agreed with this. It's confusing that a manual rebuild is needed to actually get an updated result of the test.

**Comment by jla2000** [comment]
There is https://github.com/hfn92/cmake-gtest.nvim which does achieve running Gtest tests by querying for the correct executables using cmake-tools.nvim. Maybe it could be used as a starting point on how to retrieve the required information from the cmake-tools plugin.

**Comment by mr-j0nes** [comment]
Indeed, this would be much appreciated.

**Comment by ll-nick** [comment]
For me, this is a bit of a show stopper. I work with very large code bases where this causes two problems:
- It's quite cumbersome to point to the correct executables.
- I only build tests for the parts of the code I am working on. With the current version of the repo, the test summary will be populated with hundreds of tests, most of which I do not care about. With it being so cluttered up, it's almost impossible to use. So it would be amazing, if only tests show up, where the corresponding executable is found.

Since I am very interested in a solution (everything else works flawlessly!) I'd try and give it a shot myself. This one might be a bit beyond my capabilities though, so if you could give some hints/pointers/ideas on how to approach it, that would be much appreciated.

**Comment by alfaix** [comment]
Hey @ll-nick, assuming your build tool is CMake, you can use cmake-tools.nvim to integrate; I did [this MR](https://github.com/Civitasv/cmake-tools.nvim/pull/246) some time ago and it was recently merged, which should allow you to use the lua API. You will have to convert this API to coroutines for use with nio, see how they do it [here](https://github.com/nvim-neotest/nvim-nio/blob/master/lua/nio/uv.lua#L72) via task.wrap. While I was working on it, there were a lot of rough edges in the plugin, stemming mainly from the lack of tests, so if your subset of the API ends up being small, you can use CMake API directly as a last resort.

Then you can write a couple of exploration unit tests for the functionality you would need in CMake (that is: find a target by source file; find a target by executable; recompile a target), and add those in the correct path.

When discovering executables, before prompting the user, you can query CMake first if there's a CMakeLists.txt and use that as the default mapping from files to executables. For the first iteration you can probably ignore ambiguities (if a file belongs to multiple CMake targets).

Recompiling should be the easier part as there's less logic to do; once you know the executable you're going to run, you can unconditionally recompile the relevant target. CMake does incremental compilation for you so even if no changes occurred this is quite fast.

If your tool is Bazel or Mason or anything else, I'm afraid I can't help much. I assume they have some kind of API, but I don't know of lua wrappers for them.

Lately it's very hard to find the time for this, so if you have the time and the desire to work on this I'd be more than happy to review a PR.

**Comment by ll-nick** [comment]
Sorry, forgot to mention that. Yes, CMake it is!

Thanks a lot for your response. Time is always hard to come by but I will see what I can do. 

> This issue I'm not sure I fully understand; shouldn't tests from other folders be collapsed in the summary? Unless all your tests are in the same folder I think it should be easy to navigate, unless I'm missing something.

You're right, there may be a way to filter some tests using neotest directly, I'll look into that.

For some more context: I work with ROS where you basically work with a bunch of CMake projects ("packages") in one so called workspace. Usually, you only work on a hand full of packages directly (hence are interested in the tests) but the workspaces I am working on might contain several hundred of these packages. So even if all folders are collapsed, these might still be in the triple-digits.

Setting some kind of test root directory might fix that issue though, thanks for the suggestion :)

**Comment by ll-nick** [comment]
Just wanted to give a brief update on this, maybe someone else might also be interested in moving this along. I'm still getting into the whole lua thing and it still feels like I have no clue what I am doing :sweat_smile: . Nevertheless,  I did manage to get automatic test discovery working [on my fork](https://github.com/ll-nick/neotest-gtest/tree/cmake-test-discovery) using this simple function in the `ExecutablesRegistry` class:

```lua
function ExecutablesRegistry:discover_cmake_executables()
  local model_info = cmake_tools.get_model_info()
  local build_dir = utils.normalize_path(tostring(cmake_tools.get_build_directory()))

  for _, target_info in pairs(model_info) do
    if target_info.type == "EXECUTABLE" then
      local executable_path_relative = target_info.artifacts[1].path
      local executable_path = build_dir .. lib.files.sep .. executable_path_relative
      for _, source in ipairs(target_info.sources) do
        if config.is_test_file(source.path) then
          local source_path = utils.normalize_path(source.path)
          self._node2executable[source_path] = executable_path
        end
      end
    end
  end
end
```

So I guess the next steps include making this an async function as you mentioned above and keeping up with your impressive test coverage.

**Comment by alfaix** [comment]
Hi @ll-nick , yes I think the process makes sense generally, little details can be figured out after the fact. 

One potential source of trouble to look out for is the current cmake directory; likely some switching would have to take place to allow multiple contexts (i.e., something like `old_root = cmake.get_directory(); cmake.set_directory(tests_root); do_stuff(); cmake.set_directory(old_root)`). I remember I spent some time ironing that out before giving up 😅

**Comment by AbaoFromCUG** [comment]
Just a copying of @ll-nick 's answer, as a lazy lookup executable path.

```diff

diff --git a/lua/neotest-gtest/executables/registry.lua b/lua/neotest-gtest/executables/registry.lua
index 591141e..49f46e9 100644
--- a/lua/neotest-gtest/executables/registry.lua
+++ b/lua/neotest-gtest/executables/registry.lua
@@ -2,6 +2,7 @@ local Storage = require("neotest-gtest.storage")
 local utils = require("neotest-gtest.utils")
 local lib = require("neotest.lib")
 local neotest = require("neotest")
+local config = require("neotest-gtest.config")
 
 local ADAPTER_PREFIX = "neotest-gtest:"
 
@@ -65,7 +66,9 @@ end
 function ExecutablesRegistry:find_executables(id)
   self:_reload_tree()
   self:_ensure_node_within_root(id)
-  local exe = self._node2executable[id] or self:_lookup_ancestor_executable(id)
+  local exe = self._node2executable[id]
+    or self:_lookup_ancestor_executable(id)
+    or self:_lookup_cmake_executable(id)
   if exe ~= nil then
     return { [exe] = { id } }, nil
   end
@@ -91,6 +94,33 @@ function ExecutablesRegistry:_lookup_ancestor_executable(id)
   end
 end
 
+function ExecutablesRegistry:_lookup_cmake_executable(id)
+  local success, cmake_tools = pcall(require, "cmake-tools")
+  if not success then
+    return
+  end
+
+  local ancestors = utils.collect_iterable(self:_iter_ancestors(id))
+  local ancestors_set = utils.list_to_set(ancestors)
+  local build_dir = utils.normalize_path(tostring(cmake_tools.get_build_directory()))
+  local model_info = cmake_tools.get_model_info()
+
+  for _, target_info in pairs(model_info) do
+    if target_info.type == "EXECUTABLE" then
+      local executable_path_relative = target_info.artifacts[1].path
+
+      for _, source in ipairs(target_info.sources) do
+        local source_path = utils.normalize_path(source.path)
+        if ancestors_set[source_path] then
+          -- If the source is an ancestor of the executable, then this is a test file
+          local executable_path = build_dir .. lib.files.sep .. executable_path_relative
+          return executable_path
+        end
+      end
+    end
+  end
+end
+
 function ExecutablesRegistry:_group_children_by_executable(id)
   local children_exe2nodes = {}
 

```

**Comment by haikion** [comment]
I've been testing this plugin for a week or two now. I also found it annoying to configure test binaries manually, especially when the test project consists of multiple Google Test executables.  

The submitted patch sort of works, but I encountered a few issues that needed fixing. I also added some tests. The work is included in my [fork](https://github.com/haikion/neotest-gtest/tree/executable-discovery-by-cmake) and is ready for a PR.

**Working**
- Automatic test executable discovery for a test file
- Automatic test executable discovery for a test fixture
- Automatic test executable discovery for a test

**Known issues**
- Does not automatically discover executables for a directory
- Issues may arise if the current working directory is not in the project root

I omitted the implementation for directory-based discovery, as the plugin seems to assume that a directory maps to a single test executable, which is not always the case. `cmake-tools` expects the current working directory to be in the project root, so fixing such issues would quickly become complex.

The real blocker for progress now is not technical but legal.

@AbaoFromCUG  
Can you license your patch so it can be used in a PR?

@AbaoFromCUG  
@ll-nick  
Let me know if you want your nick (or name) mentioned in the comments. If not, then please use a permissive license such as [Unlicense](https://unlicense.org).

@alfaix 
Let me know your views on opening a PR.

**Comment by AbaoFromCUG** [comment]
@haikion Feel free to create a PR if this helps 👍

**Comment by ll-nick** [comment]
I'm not doing as much C++ anymore currently that's why this issue got left behind. 

If you have a solution, go for it. Feel absolutely free to use the snippet I posted above, that was the intention after all. You may mention me if that makes sense somehow but there is no need.

**Comment by haikion** [comment]
Thank you for the permissions. I've created a PR.
